### PR TITLE
add 'linearize_sources' API to `StorageController`

### PR DESCRIPTION
**Pull request by guswynn**
This PR adds the `LinearizeSources` and `LinearizedTimestamps` variants to `StorageCommand` and `StorageResponse` respectively. This is the main api boundary between the coordinator and the `STORAGE` layer that allows the coordinator, once the logic is implemented, to "wait" on storage to catch a source up to a specific set of timestamps, giving the illusion of linearized reads for certain scenarios (and up-to-date ones for others).

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

This PR intentionally leaves out a thorough implementation, as its intended to iterate, through review, on the storage API itself

### Testing

- [N/A] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

None

### CODE REVIEW

File: src/dataflow-types/src/client.rs [hunk 1]
```diff
@@ -1111,6 +1134,10 @@ pub mod partitioned {
                         feedback,
                     ))))
                 }
+                // TODO(guswynn): is this the correct implementation?
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client.rs:1137]: It's hard to know without the implementation. If each worker collects the max offsets for the parts that it manages, then we want to integrate the results before returning them.

Comment (guswynn) [reply src/dataflow-types/src/client.rs:1137]: I think that each _storage_ worker will just produce the same results?

**Review by benesch — APPROVED** [review]
This all checks out for me. We should get someone from STORAGE to give the official signoff, though! Either @petrosagg or someone he nominates?

**Review by frankmcsherry — COMMENTED** [review]
I think the API looks different from what I would imagine, as I don't know that you can use the one specified here to get the right result. Specifically, 
1. I think `linearize_sources` needs to take an additional `peek_id` argument, or something that uniquely identifies the request.
2. I think that the response should just be `(PeekId, Timestamp)`, telling you the minimum single timestamp appropriate for the sources you've asked about.

With these changes, I could see how you could use the methods to get the desired behavior. Without the changes, I'm not clear on how a person trying to determine a timestamp for a peek will be able to ask, and get a certain result back for their peek.

Does this sound right?

**Comment by benesch** [comment]
Ah, (1) is a very good point. Is (2) mostly a question of who does the `min` across all the partitions (STORAGE vs ADAPTER)?

**Comment by benesch** [comment]
Makes sense to me! I just wanted to make sure that (1) was the actual
correctness issue, and (2) was (at least somewhat) a matter of taste.

File: src/dataflow-types/src/client/controller/storage.rs [hunk 2]
```diff
@@ -99,6 +100,13 @@ pub trait StorageController: Debug + Send {
         updates: &mut BTreeMap<GlobalId, ChangeBatch<Self::Timestamp>>,
     );
 
+    /// Send a request to obtain "linearized" timestamps for the given sources.
+    async fn linearize_sources(
+        &mut self,
+        peek_id: Uuid,
+        source_ids: Vec<GlobalId>,
+    ) -> Result<(), anyhow::Error>;
```
Comment (petrosagg) [inline src/dataflow-types/src/client/controller/storage.rs:108]: @frankmcsherry @benesch I was wondering, why do we go for this pattern where these async methods fire a command to storage and then the caller needs to observe the response stream to figure out what happened?

It sounds to me that this function interface would be much better for users interacting with STORAGE:

```rust
async fn linearize_sources(&mut self, source_ids: Vec<GlobalId>) -> Result<Self::Timestamp, anyhow::Error>;
```

i.e you get back a `Future` that will resolve to your result and there is no need to specify a `peek_id`. It's the responsibility of the storage controller implementation to figure out a way to match a request to a response and you don't have to dissect the `recv` stream to figure out what happened. 

Do we lose something if we design it that way?

Comment (frankmcsherry) [reply src/dataflow-types/src/client/controller/storage.rs:108]: I'm not sure! We potentially lose some visibility of the command and response flow (currently `trace!`d output). I find that very helpful for the compute commands (e.g. why not do the same for `Peek`?) but I could also imagine that we could keep that behaviour and wrap it up differently if we want.

There is a more complicated multi-way `select!` to write I think, if we have 1k outstanding requests of this form, that a channel addresses nicely.

My personal current feelings are that `async` is a grim antipattern and the less we do of it the better (for reasons of cancelation safety), but that may just be due to being burned by it recently. Putting things in a queue is much more clearly not-wrong, for me at least.

**Review by petrosagg — APPROVED** [review]

File: src/dataflow-types/src/client/controller/storage.rs (hunk 2 above)
Comment (guswynn) [reply src/dataflow-types/src/client/controller/storage.rs:108]: @petrosagg In the work I am doing that follows this, `linearize_sources` is async, but all it does it `spawn` a tokio task  that will eventually send a response back on the channel back to the coordinator. In that sense, this method doesn't _really_ need to be async, and we can simplify it later if need be. If we did all the work inline in this `async fn`, then we would hang the coordinator waiting for sources to catch up to the linearized offsets

**Pull request merged**